Improve the performance of cluster chain allocation

Konamiman opened the pull request:
On FAT16 drives that have 64K clusters and are almost full, allocating new clusters is a very slow process. The entire FAT is scanned every time, and copying big files, a process that involves many allocations, takes an absurd amount of time. As an user mentioned in [the msx.org forum](https://msx.org/forum/msx-talk/development/my-plans-for-nextor-v3-warning-broken-drivers-ahead):

> I once tested with 4Mb partition filled 75% and copying 16Mb file into it - after one hour I just turned computer off - it was still copying stuff

This pull request introduces a mechanism that dramatically improves the performance of the cluster chain allocation process: a per-drive cache of the last cluster number allocated that prevents the search process from starting over at the beginning of the FAT every time.

## How is it implemented

A new field named `UD_ACLU` is added to the unit descriptor (a data structure that exists for each mapped drive). This field stores the cluster number where the search for free clusters will start the next time a cluster chain needs to be allocated (minus one). It's initialized to 1 when the descriptor is created (at boot time, or when the drive is explicitly mapped at a later time) and modified as follows:

- After a cluster chain allocation process succeeds, it's set to the number of the last cluster that has been allocated.

- When a cluster chain is freed, it's set to the number of the lowest cluster number from the chain that has been freed, minus one, but **only** if that new value would be lower than the previous value.

This algorithm guarantees that the allocation process works as usual (the lowest numbered free clusters are allocated first), but much faster.

Additionally, `UD_ACLU` is initialized to 1 again if:

- The allocation of a cluster chain fails with a "Disk full" error (or with any other error, actually, but "Disk full" will be the most common case). This is a fail safe mechanism: if the algorithm has any flaw and trying to write to the drive causes a "Disk full" error even though there is actually some free space available, the next attempt will succeed.

- The drive is written to by using the raw sector write `RDABS` or `WRDRV` function calls. This is to account for the case where a disk optimizer tool writes to the FAT sectors directly; the number of the sectors being written is not checked to keep the code simple (and direct sector writing using these functions is an uncommon case after all).

## How to test

The easiest way to test is to use an emulator.

1. Create a disk image with a size that is a power of 2, so that it gets the maximum number of clusters (close to 64K). A size of 128MB will be enough.

2. Fill the disk image with files until it's almost completely full (leave e.g. about 100K free).

3. Boot with that disk image, go to BASIC and create a trivial program e.g. `10 REM`. Save it, `save"x"`. It will take about 5s for a R800 or about 30s for a Z80.

4. Save the program again with a different name, notice how the process is now immediate.

If you want to try the "copy a 16MB file" scenario instead, you'll see that now a R800 needs about two minutes to complete the process, which is about the same time as when the drive is empty.

Konamiman committed:
Improve cluster allocation by caching the last allocated cluster number.

On FAT16 drives that have 64K clusters and are almost full, allocating
new clusters is a very slow process. The entire FAT is scanned every
time, and copying big files, a process that involves many allocations,
takes an absurd amount of time.

This commit introduces a mechanism that dramatically improves the
performance of the process:

- A new field named UD_ACLU is added to the unit descriptor.
  This field is initialized to 1 when the descriptor is created.

- Whenever a cluster chain allocation happens, the process of searching
  free clusters doesn't start at cluster 2, instead it starts at the
  value stored at UD_ACLU plus one.

- If the allocation process succeeds, UD_ACLU is updated to the number
  of the last cluster that has been allocated.

- When a cluster chain is freed, UD_ACLU is updated to the number
  of the first cluster in the chain minus one, but ONLY if that
  new value is lower than the previous value.

This algorithm guarantees that the allocation process works as usual
(the lowest numbered free clusters are allocated first), but
much faster.

Additionally, UD_ACLU is initialized to 1 again if:

- The drive is written to by using the raw sector write RDABS or WRDRV
  function calls.

- The allocation of a cluster chain fails (typically with a "disk full"
  error).

Pencioner commented on source/kernel/bank2/fat.mac:
```diff
@@ -915,6 +945,45 @@ DTY_ADD:
+; Corrupts:    None
+;
+ACLU_GET:
+	push ix	
+	push hl
+	pop ix
+	ld e,(ix+UD_ACLU##)
+	ld d,(ix+UD_ACLU##+1)
+	pop ix
```
I would go with

    push hl 
    push bc 
    ld bc, UD_ACLU## 
    add hl, bc  
    ld e, (hl)  
    inc hl  
    ld d, (hl)  
    pop bc 
    pop hl 

I had calculated bytes and cycles (including that famous one extra wait for every command) and if i'm not mistaken, avoiding use of IX would make code 2 bytes shorted and 17 cycles faster. I guess ACLU_SET would benefit from similar change too

Pencioner replied:
Well, missed that my version will corrupt AF unfortunately. Sorry for noise

Konamiman replied:
What if we `push af` and `pop af` then? The code wouldn't be shorter but we would still gain some cycles, right?

theNestruo replied:
What about using DE instead of BC in @Pencioner version?:

```
	push	af
	push	hl
	ld	de, UD_ACLU##
	add	hl, de
	ld	e, (hl)
	inc	hl
	ld	d, (hl)
	pop	hl
	pop	af
; 92 t-states (-11), 11 bytes (-2)
```

If I'm not mistaken:
- Original code: 103 t-states, 13 bytes
- Pencioner proposal: 92 t-states (-11), 11 bytes (-2) _(but corrupts F, as stated)_
- Pencioner proposal with `push af` and `pop af`: 115 t-states (+12), 13 bytes
- Pencioner proposal using DE: 92 t-states (-11), 11 bytes (-2)

I'm not familiar with Nextor code, so I don't know if it is required not to corrupt F... If the `push af` and `pop af` can be removed, the routine is: 69 t-states (-34), 9 bytes (-4)

Konamiman replied:
`ACLU_GET` is called in two places, one needs AF to be preserved and the other doesn't, so I guess the responsibility of preserving AF could be moved to the caller. And for sure a similar optimization could be done for `ACLU_SET` too; I'll take a look, thank you very much!

theNestruo replied:
Moving the responsability of preserving AF to the callers sounds good.

Are these those two places?
- https://github.com/Konamiman/Nextor/blob/improved-free-cluster-search/source/kernel/bank2/fat.mac#L656
- https://github.com/Konamiman/Nextor/blob/improved-free-cluster-search/source/kernel/bank2/fat.mac#L891

I'm trying to find reads of A or the flags before being overwritten after calling ACLU_GET and I can't find them...

Konamiman replied:
Yep, these two. And for the case of `AL_CLUSTERS`, a parameter is passed in A but indeed it's stored for later usage before `ACLU_GET` is called, so looks like it's not actually necessary to preserve it or the flags.

theNestruo commented on source/kernel/bank2/fat.mac:
```diff
@@ -869,6 +880,25 @@ zero_dsec_loop:	ld	(hl),c			; the data area of buffer
+		;This is to ensure that UD_ACLU is always either 1 or the
+		;lower free cluster number available -1.
+
+		push de
+		pop bc			;BC = first cluster to free (new value+1)
+		push hl
+		call ACLU_GET	;DE = current value of UD_ACLU
+		push bc
+		pop hl          ;HL = first cluster to free (new value+1)
+		ex de,hl        ;HL = current value, DE = new value+1
+		dec de          ;DE = new value
+		or a
+		sbc hl,de       ;HL = current value - new value, Cy=1 if new > current
+		pop hl          ;HL = address of unit descriptor
```
I think these lines can be optimized as:
```
		push de		;SP = First cluster number in chain
		call ACLU_GET	;DE = current value of UD_ACLU
		ex (sp), hl     ;HL = first cluster to free (new value+1), SP = Address of unit descriptor
		ex de,hl        ;HL = current value, DE = new value+1
		dec de          ;DE = new value
		or a
		sbc hl,de       ;HL = current value - new value, Cy=1 if new > current
		pop hl          ;HL = address of unit descriptor
```
Gaining 26 t-states and 3 bytes.

Konamiman replied:
For some reason my brain has a difficult time handling the `ex (sp),hl` tricks so I don't use these often, but that optimization looks great, thanks!

Konamiman committed:
Code optimizations in ACLU_*ET and FR_CHAIN

theNestruo commented on source/kernel/bank2/rw.mac:
```diff
@@ -153,6 +153,12 @@ abs_common_doerr:
 		jr	try_rw_val
 
 abs_rw_dorw:
+		push de		;If writing, reset number of next cluster to allocate
+		bit	RF_READ,(iy+@RW_FLAGS##) ;just in case we're messing up with FAT
+		ld de,1
+		pcall z,ACLU_SET
+		pop de
```
If you can spare 9 bytes, I would suggest:
```
    bit RF_READ,(iy+@RW_FLAGS##) ;just in case we're messing up with FAT
    pcall z,ACLU_RESET ;If writing, reset number of next cluster to allocate
```

And then implement `ACLU_RESET` as:
```
	PROC	ACLU_RESET
;
;    Resets the UD_ACLU field of a unit descriptor (sets to -1)
;
; Entry:   HL = Address of unit descriptor
; Returns:     None
; Corrupts:    F
;
	push	hl
	push    de
	ld	de, UD_ACLU##
	add	hl,de
	ld	(hl),0ffh
	inc	hl
	ld	(hl),0ffh
	pop     de
	pop	hl
	ret
```

If I'm not mistaken, this is 28 t-states faster if the branch is taken (149 vs 177 t-states) and 34 t-states faster otherwise (33 vs 67 t-states)

Konamiman replied:
Thanks for the suggestion but I think I'll leave this one unchanged. ROM space is becoming scarce so the smaller the code the better, and resetting `UD_ACLU` is something that won't happen often, so it's a process that doesn't need to be heavily optimized.

Konamiman committed:
Fix bad line endings in val.mac

Konamiman committed:
Fix the update of UD_ACLU in FR_CHAIN.

FR_CHAIN was simply updating UD_ACLU to the first cluster number
in the freed chain, but this was wrongly assuming that cluster numbers
in a FAT chain always increase, while they could actually decrease
(e.g. the chain could be 200-201-202-100-FFF).

The fix implemented in this commit uses FAT_CHAIN, a variable used only
by AL_CLUSTERS, to temporarliy keep track of which is the lowest
cluster number freed during the process, then uses this value instead
of the first cluster number in the chain to update UD_ACLU
if appropriate.

Konamiman commented:
msx.org user pgimeno [appropriately comments](https://msx.org/forum/msx-talk/development/fast-cluster-allocation-in-nextor-reviewerstesters-needed):

>  It's perfectly possible that a cluster chain contains a cluster with a number lower than the first one in the chain.

@Pencioner @theNestruo I've pushed a commit that updates `FR_CHAIN` to take this in account, could you please take a look?

theNestruo commented:
_Edit: saved 3 more bytes._

_Disclaimer: untested code._

Minor optimization (1 byte, 3 t-states) of frchlp2 #916-924:

```
	push	de		; Save cluster number
	call	FAT_LOOKUP	; Get the next cluster number in the chain
	ex	de, hl		; hl = next cluster number, de = old hl
	ex	(sp), hl	; hl = saved cluster number, (sp) = next cluster number
	ex	de, hl		; hl = old hl, de = saved cluster number
	ld	bc, 0		; Set FAT entry for this cluster to 0 to free it
	call	FAT_?SET
	pop	de		; de = next cluster number
```

Also, I think the jump in #888, `ld (FAT_CHAIN##),de / call free_chain_lp`, can jump to `frchlp2 + 1` (after the `pop hl`). `free_chain_lp`, in this first iteration, is checking `FAT_CHAIN##` against `FAT_CHAIN##`... so we can skip that iteration and save 97/92 t-states.
Swapping the labels (so `free_chain_lp` is still the entry point) and moving the `pop hl` a few lines up:
_Edit:_ And we can use the `ld (FAT_CHAIN##),de` of `free_chain_lp` and remove it from #888, saving 3 more bytes:

```
frchlp2:
	push hl			;Preserve address of unit descriptor
	ld hl, (FAT_CHAIN##)
	or a
	sbc hl,de		;HL = last cluster freed - cluster to free 
	pop hl			;Restore address of unit descriptor ; (moved up)
	jr c, frchlp3		;Cy = 1 means last cluster freed < cluster to free
free_chain_lp:
	ld (FAT_CHAIN##),de
frchlp3:
	push	de		; Save cluster number
	call	FAT_LOOKUP	; Get the next cluster number in the chain
	ex	de, hl		; hl = next cluster number, de = old hl
	ex	(sp), hl	; hl = saved cluster number, (sp) = next cluster number
	ex	de, hl		; hl = old hl, de = saved cluster number
	ld	bc, 0		; Set FAT entry for this cluster to 0 to free it
	call	FAT_?SET
	pop	de		; de = next cluster number
	ld	a,d		; back and return if it is
	or	e		; zero or -ve (end of chain).
	ret	z
;===== start mod FAT16 (CLST_7 patch)
	ld	a,d
	and	e
	inc	a
	jr	nz,frchlp2	;Loop if not end of chain.
;===== end mod FAT16
	ret
```

Konamiman committed:
Code optimization for the FR_CHAIN routine.

Konamiman commented:
@theNestruo Optimization implemented. And with that one I think we're all set.
